juniper/src/types/marker.rs

```diff
@@ -11,8 +16,108 @@ use crate::{GraphQLType, ScalarValue};
 /// GraphQL objects. Other types (scalars, enums, and input objects)
 /// are not allowed.
 pub trait GraphQLObjectType<S: ScalarValue>: GraphQLType<S> {
-    #[allow(missing_docs)]
-    fn name(info: &Self::TypeInfo) -> Option<&str> {
-        <Self as GraphQLType<S>>::name(info)
-    }
+    /// An arbitrary function without meaning.
+    ///
+    /// May contain compile timed check logic which ensures that types
+    /// are used correctly according to the GraphQL specification.
+    fn mark() {}
```

Comment:
I believe this can be empty and might slightly save on compile times. At least, the implementation of traits on https://doc.rust-lang.org/std/marker/index.html are empty. Could be done in a followup.

Reply:
The problem is that I did not found any other way of type checking except for `<T as U>::mark()`. Therefore, we require at least one function inside the trait to ensure that `T` implements `U`. Adding the trait bound as `where` clauses introduces a large overhead in the implementation and would require to specify lifetimes in struct like `Test<'a>`. In most internal structs lifetimes bound inside the struct can be ignored with the current approach.